### PR TITLE
Transform destructuring private

### LINKED ISSUE #13889: Support destructuring-private
### 💻

- [X] Would you like to work on this feature?

### What problem are you trying to solve?

[Destructuring Private](https://github.com/tc39-transfer/proposal-destructuring-private) proposal just reached Stage 2. Babel should support parsing as well as transforming this feature.

### Describe the solution you'd like

For parsing, we can add a new parser plugin `DestructuringPrivate` and extend `parseObjectProperty` to cover new syntax.

For transforming, we should only transform the private destructuring. For example,

```js
// In
const { #x: x, y = x, #z: { #w: w, ...v } } = o;

// Out
const x = o.#x;
const { y = x } = o;
const o$z = o.#z;
const w = o$z.#w;
const { ...v } = o$z; // private fields are not enumerable and thus `v` should not contain `#z`
```

### Describe alternatives you've considered

For transforming, we can also run the full destructuring transform, but the output will be more verbose.

### Documentation, Adoption, Migration Strategy

_No response_

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Transforms [destructuring-private](https://github.com/tc39/proposal-destructuring-private), closes #13889
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | https://github.com/babel/website/pull/2631
| Any Dependency Changes?  |
| License                  | MIT

In this PR we implement [destructuring-private](https://github.com/tc39/proposal-destructuring-private) transform. The design goal is:

1. Transpile as little as possible
2. Ensure the execution order

The PR is based on #14236. In #14236 we refactor [destructuring-transform](https://github.com/babel/babel/blob/main/packages/babel-plugin-transform-destructuring/src/index.ts) a bit so that some routines can be reused by the this plugin.

## Technical details
The plugin is composed of two parts, `index.ts` and `util.ts`.

The first part converts destructuring in function params / catch clause / forXstatement `for (const { #x: x } of arr);` to variable declarations / assignment expressions. This part is largely based on current [destructuring-transform](https://github.com/babel/babel/blob/main/packages/babel-plugin-transform-destructuring/src/index.ts). Note that unlike the general destructuring transform, we don't handle the export declarations because private destructuring must be placed inside a class while an export declaration must be in top level.

The second part tranpiles private destructuring in variable declarations / assignment expressions.

We first implement a simplified generator-based DFS traverser, which traverses given pattern and ignores expressions within computed keys and initializers. The traverser will be paused when it finds a private key, and reports an index path.

Then we split given pattern around the private keys. The reported index path is handled from left to right: Any patterns before private key is guaranteed to be private-key free and thus yielded to the caller. Any patterns after private key is uncharted territory and pushed to a stack so we will revisit later. The destructuring private key are then transformed and its `value`, which may be a new pattern, are also pushed to a stack. The transform ends when the stack is drained.

#### Object patterns
```js
var { [f()]: a, #x: { #y: y }, [g()]: b } = this;
```
are split into
```js
var { [f()]: a } = this,
{ #y: y } = this.#x,
{ [g()]: b } = this;
```

#### Object patterns with rest element
```js
var { [f()]: a, #x: { #y: y }, [g()]: b, ...c } = this;
```
are split into
```js
var m1;
var { [m1 = f()]: a } = this,
{ #y: y } = this.#x,
{ [g()]: b, ...c } = babelHelpers.objectWithoutProperties(this, [m1]);
```
This approach is a bit deviated from how we handle non-static property keys in the general transform: In that plugin all non-static property keys are memoised before the pattern, which is not spec-compliant because computed keys should be evaluated right after the previous pattern elements are handled. See the ordering fixtures test for more examples.

#### Array patterns
```js
var [{ #x: x = f() }, y] = iterator();
```
are transformed into
```js
var [m1, m2] = iterator(),
  { #x: x = f() } = m1,
  y = m2;
```

This approach assumes that any expressions within the array pattern, i.e.`f()`, does not interfere with the iterable `iterator()`. This assumption is generally held as array destructuring is meant to consume an iterator, and if one mutates the iterable in the middle, then one had better use a loop statement instead to convey such purposes. To be complete spec-compliant, we will have to pause the iterator step by step and interpolates transpiled destructuring codes, like @nicolo-ribaudo mentioned in [a gist](https://gist.github.com/nicolo-ribaudo/f8ac7916f89450f2ead77d99855b2098). We didn't implement such transform as it is bloated and the general destructuring transform has not supported this case yet.

## Playground
You can test the behaviour in the REPL mentioned by babel-bot below. You have to enable the `stage-2` preset, currently the REPl can not disable class properties transforms as they are also included in `stage-2` preset.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14304"></a>

### CODE REVIEW

File: packages/babel-plugin-proposal-destructuring-private/test/fixtures/ordering/.object-pattern-with-rest/exec.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    var { [push(0)]: a = push(1), #x: {
+      [push(3)]: b = push(4),
+      #y: y = push(5),
+      [push(6)]: c = push(7)
+    } = push(2, C), [push(8)]: d = push(9), ...e } = C;
+  }
+}
+
+var nums = Array.from({ length: 10 }, (_, i) => i);
+expect(log).toEqual(nums);
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-destructuring-private/test/fixtures/ordering/.object-pattern-with-rest/exec.js:18]: This test is disabled because `plugin-transform` evaluates `push(8)` immediately after `push(0)`, which should be fixed in a separate PR.

File: packages/babel-plugin-proposal-destructuring-private/test/fixtures/typescript/.valid-preset-typescript/options.json [hunk 2]
```diff
@@ -0,0 +1,4 @@
+{
+  "plugins": ["proposal-destructuring-private"],
+  "presets": ["typescript"]
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-destructuring-private/test/fixtures/typescript/.valid-preset-typescript/options.json:3]: This test is disabled because `presets` always run _after_ `plugins`. The order is preferred for preset-env and stage plugins, but not for typescript preset since it should run before other es plugins.

File: packages/babel-plugin-proposal-destructuring-private/test/fixtures/assignment--es2015/array-rest/output.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+var _x = {
+  writable: true,
+  value: void 0
+};
+
+(() => {
+  var _m, _p, _p2, _m2;
+
+  _m = [C], [_p, ..._p2] = _m, _m2 = babelHelpers.classStaticPrivateFieldSpecGet(_p, C, _x), x = _m2 === void 0 ? 1 : _m2, z = _p2;
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-destructuring-private/test/fixtures/assignment--es2015/array-rest/output.js:13]: `_m` is not needed here, maybe we could split the assignment only if we need to reference it multiple times.

File: packages/babel-plugin-proposal-destructuring-private/package.json [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      null,
+      {
+        "exports": null
+      }
+    ]
+  },
+  "exports": {
+    ".": "./lib/index.js",
+    "./package.json": "./package.json"
+  }
```
[the remark this answers is not in the document]
Comment (JLHwung) [reply packages/babel-plugin-proposal-destructuring-private/package.json:49]: I think they were like you suggested but then some yarn plugins complains and the e2e-publish test failed, I don't remember the exact errors, I will commit the suggestion and see what will happen.

File: packages/babel-plugin-transform-destructuring/src/index.ts [hunk 5]
```diff
@@ -80,7 +80,9 @@ export default declare((api, options) => {
           }
 
           statementBody.unshift(
-            t.expressionStatement(t.assignmentExpression("=", left, temp)),
+            t.expressionStatement(
+              t.assignmentExpression("=", left, t.cloneNode(temp)),
```
Comment (JLHwung) [inline packages/babel-plugin-transform-destructuring/src/index.ts:84]: `temp` has been reused when creating `node.left`. I am surprised that `babel-check-duplicated-nodes` does not report this issue. However, it reports when I happen to delete
```
if (statementBody.length === 0 && path.isCompletionRecord()) {
            statementBody.unshift(
              t.expressionStatement(scope.buildUndefinedNode()),
            );
          }
```

File: packages/babel-plugin-proposal-destructuring-private/src/util.ts [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+            restExcludingKeys,
+            right,
+            scope,
+            addHelper,
+            objectRestNoSymbols,
+            useBuiltIns,
+          ),
+        };
+      } else {
+        // @ts-expect-error left must not contain RestElement if restExcludingKeys is nullish
```
[the remark this answers is not in the document]
Comment (JLHwung) [reply packages/babel-plugin-proposal-destructuring-private/src/util.ts]: Oh the wording is correct, it should be `left` must not be a `RestElement` so it is okay to yield `{left, right}`. I have tuned the types so this comment is not required.

File: packages/babel-plugin-proposal-destructuring-private/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+      scope.crawl();
+    },
+
+    AssignmentExpression(path, state) {
+      const { node, scope, parent } = path;
+      if (!hasPrivateKeys(node.left)) return;
+      const assignments = [];
+      const shouldPreserveCompletion =
+        (!isExpressionStatement(parent) && !isSequenceExpression(parent)) ||
+        path.isCompletionRecord();
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-destructuring-private/src/index.ts:145]: Why do we exclude `ExpressionStatement` and `SequenceExpression`?

Comment (JLHwung) [reply packages/babel-plugin-proposal-destructuring-private/src/index.ts:145]: It is copied from
https://github.com/babel/babel/blob/9cb5f6c983b6d4be94de880946c16970a3788c9f/packages/babel-plugin-transform-destructuring/src/util.ts#L666

`NodePath#isCompletionRecord` essentially checks alongside the ancestry if the path is the last item of its container. This approach works for ExpressionStatement in Block and expression in SequenceExpression.

**Review by nicolo-ribaudo — APPROVED** [review]

File: packages/babel-plugin-proposal-destructuring-private/test/fixtures/ordering/.array-pattern-side-effect-iterable/exec.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+var log = [];
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-destructuring-private/test/fixtures/ordering/.array-pattern-side-effect-iterable/exec.js:1]: This test is disabled because currently the plugin does not support it: in `transformPrivateKeyDestructuring` the sub patterns of the array pattern as LHS is evaluated after the iterable at RHS is consumed.

File: packages/babel-plugin-proposal-destructuring-private/src/util.ts [hunk 9]
```diff
@@ -257,12 +263,16 @@ export function* privateKeyPathIterator(pattern: t.LVal) {
   });
 }
 
+// t.LVal without t.RestElement
 type LHS =
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-proposal-destructuring-private/src/util.ts:267]: You could try using https://www.typescriptlang.org/docs/handbook/utility-types.html#excludeuniontype-excludedmembers

File: packages/babel-plugin-proposal-destructuring-private/src/util.ts [hunk 10]
```diff
@@ -0,0 +1,465 @@
...
+ * @param {t.LVal} pattern
+ */
+export function* privateKeyPathIterator(pattern: t.LVal) {
+  const indexPath = [];
+  yield* traversePattern(pattern, function* (node, index, depth) {
+    indexPath[depth] = index;
+    if (isObjectProperty(node)) {
+      const propertyKey = node.key;
+      if (isPrivateName(propertyKey)) {
+        yield indexPath.slice(1, depth + 1);
```
Comment (jridgewell) [inline packages/babel-plugin-proposal-destructuring-private/src/util.ts:254]: Can you add a comment why we're doing `1` and `depth + 1`?

File: packages/babel-plugin-proposal-destructuring-private/src/util.ts [hunk 11]
```diff
@@ -0,0 +1,465 @@
...
+    right,
+    restExcludingKeys: initRestExcludingKeys(left),
+  });
+  let item: Item;
+  while ((item = stack.pop()) !== undefined) {
+    const { restExcludingKeys } = item;
+    let { left, right } = item;
+    const searchPrivateKey = privateKeyPathIterator(left).next();
+    if (searchPrivateKey.done) {
+      if (restExcludingKeys?.length > 0) {
```
Comment (jridgewell) [inline packages/babel-plugin-proposal-destructuring-private/src/util.ts:321]: Is this handling something like the following? Turning `{ first, #priv: priv, ...rest }` into `{ first }, priv = #priv, { …rest }` would break the `…rest`'s exclude keys. So wee need to handle the case where something has been excluded. But in the case of `{ #priv: priv, first, …rest }`, it doesn't matter because `priv = #priv, { first, …rest }` still has the correct behavior.

File: packages/babel-plugin-proposal-destructuring-private/src/util.ts [hunk 12]
```diff
@@ -0,0 +1,465 @@
...
+          right,
+        };
+      }
+    } else {
+      // now we need to split according to the indexPath;
+      const indexPath = searchPrivateKey.value;
+      let index;
+      let isFirst = true;
+      while (
+        (index = indexPath.shift()) !== undefined ||
```
Comment (jridgewell) [inline packages/babel-plugin-proposal-destructuring-private/src/util.ts:357]: Nit: This is n^2 array accesses.

**Review by jridgewell — APPROVED** [review]

File: packages/babel-traverse/test/scope.js [hunk 13]
```diff
@@ -902,5 +902,30 @@ describe("scope", () => {
       `);
       expect(program.scope.hasOwnBinding("class")).toBe(true);
     });
+    it("registers the new binding outside function when the path is a param initializer", () => {
+      const program = getPath("(a = f()) => {}");
+      const assignmentPattern = program.get("body.0.expression.params.0");
+      assignmentPattern.scope.push({ id: t.identifier("ref") });
+      expect(program.toString()).toMatchInlineSnapshot(`
+        "var ref;
```
[the remark this answers is not in the document]
Comment (JLHwung) [reply packages/babel-traverse/test/scope.js:910]: `var ref` is created by `assignmentPattern.scope.push({ id: t.identifier("ref") });`, this test is to show that the `ref` is pushed to the parent scope instead of
```js
class C {
  m(a = f()) { var ref; }
}
```

File: packages/babel-plugin-proposal-destructuring-private/src/index.ts [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+      const { params: transformedParams, variableDeclaration } =
+        buildVariableDeclarationFromParams(paramsAfterIndex, scope);
+
+      path.get("body").unshiftContainer("body", variableDeclaration);
+      params.push(...transformedParams);
+      // preserve function.length
+      // (b, p1) => {}
+      // transforms to
+      // (b, p1 = void 0) => {}
+      if (firstAssignmentPatternIndex >= firstPrivateIndex) {
```
[the remark this answers is not in the document]
Comment (JLHwung) [reply packages/babel-plugin-proposal-destructuring-private/src/index.ts:60]: We can not assign the `firstAssignmentPatternIndex` here because the transform will convert any params after the first private destructuring. In other words, `firstAssignmentPatternIndex` will always be less than `firstPrivateIndex` if we assign `firstAssignmentPatternIndex` here.

File: packages/babel-plugin-proposal-destructuring-private/src/util.ts (hunk 11 above)
Comment (JLHwung) [reply packages/babel-plugin-proposal-destructuring-private/src/util.ts:321]: Yes. The `growRestExcludingKeys` will exclude private keys:

https://github.com/babel/babel/pull/14304/files#diff-5742fd1087b16ca97d1e7d8fbf176f27ad52d323f5e281db9bc349bd257efd26R68

> But in the case of { #priv: priv, first, …rest }, it doesn't matter because priv = #priv, { first, …rest } still has the correct behavior.

Yes. The plugin should already handle this case, I will add new tests.

File: packages/babel-plugin-proposal-destructuring-private/src/util.ts (hunk 12 above)
Comment (JLHwung) [reply packages/babel-plugin-proposal-destructuring-private/src/util.ts]: Fixed by iterating indexPath instead.

File: packages/babel-plugin-proposal-destructuring-private/src/util.ts [hunk 15]
```diff
@@ -0,0 +1,470 @@
...
+    }
+  }
+}
+
+export function hasPrivateKeys(pattern: t.LVal) {
+  let result = false;
+  traversePattern(pattern, function* (node) {
+    if (isObjectProperty(node) && isPrivateName(node.key)) {
+      result = true;
+      yield;
```
Comment (jridgewell) [inline packages/babel-plugin-proposal-destructuring-private/src/util.ts:227]: I believe this is now just preventing `traversePattern`'s `while` loop from continuing. Can you add a comment?

**Review by jridgewell — APPROVED** [review]
